### PR TITLE
[master] Allow the resizing of virtual machine disks on OpenNebula.

### PR DESCRIPTION
**Pull request by KarlAustin**
### What does this PR do?
Allows the resizing of virtual machine disks on OpenNebula

### What issues does this PR fix or reference?
Fixes: N/A

### New Behavior
As above, allows virtual machine disks to be resized when using the OpenNebula Salt Cloud module.

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
- [x ] Docs
- [ x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [ x] Tests written

### Commits signed with GPG?
No

### CODE REVIEW

File: salt/cloud/clouds/opennebula.py [hunk 1]
```diff
@@ -3177,6 +3177,75 @@ def vm_disk_save(name, kwargs=None, call=None):
 
     return data
 
+def vm_disk_resize(name, kwargs=None, call=None):
+    """
+    Changes the capacity of the virtual machine disk.
+
+    .. versionadded:: TBD
```
Comment (Ch3LL) [inline salt/cloud/clouds/opennebula.py:3184]:
```suggestion
    .. versionadded:: 3007.0
```

File: tests/pytests/unit/cloud/clouds/test_opennebula.py [hunk 2]
```diff
@@ -1261,6 +1261,32 @@ def test_vm_detach_nic_no_nic_id():
...
+    pytest.raises(
+        SaltCloudSystemExit,
+        opennebula.vm_disk_resize,
+        VM_NAME,
+        call="action",
+        kwargs={"disk_size": "1024"},
+    )
+
+
+def test_vm_disk_resize_no_disk_size():
```
Comment (Ch3LL) [inline tests/pytests/unit/cloud/clouds/test_opennebula.py:1277]: Can we also add test coverage when all args expected are passed and the function works as we intend?

**Review by Ch3LL — CHANGES_REQUESTED** [review]
This will also require a changelog: https://docs.saltproject.io/en/latest/topics/development/changelog.html

**Comment by KarlAustin** [comment]
> Can we also add test coverage when all args expected are passed and the function works as we intend?

I guess the issue there is mocking the OpenNebula API?  There's already a test skipped at line 625 due to the mocking issue: https://github.com/saltstack/salt/blob/560bacdaeb37d87ca970a24da51c59d331ec6400/tests/pytests/unit/cloud/clouds/test_opennebula.py#L625

I've written an outline of a test for it based on the test_image_clone_success

> Changelog

Done

> Version

Done

**Comment by KarlAustin** [comment]
My initial thoughts were that I'm not sure it's easily possible without re-factoring the OpenNebula module itself.  If you look at it compared to the OpenStack module, that takes in a connection for each call - so a mocked connection can be passed in when testing.  The OpenNebula module doesn't work like that.

However, I did find a way around this.  It isn't exactly pretty, but it does work, and there may well be a better way of doing it.  I'll be honest, I'm not all that familiar with Python, or the Salt codebase.

```
"""
Mock the server object
"""
class MockServer:
    def __init__(self, ret):
        self.ret = ret
        self.one = MockOne(self.ret)


"""
Mock the one API instance
"""
class MockOne:
    def __init__(self, ret):
        self.ret = ret
        self.vm = MockObject(self.ret)


"""
Mock the object we are working on e.g. vm in the call to vm.diskresize
"""
class MockObject:
    def __init__(self, ret):
        self.ret = ret

    def __getattr__(self, name):
        return MockCallable(self.ret)


"""
Mock the callable e.g. diskresize in vm.diskresize
"""
class MockCallable:
    def __init__(self, ret):
        self.ret = ret

    def __call__(self, *args):
        return self.ret


def test_vm_disk_resize_success():
    """
    Tests that disk_resize returns successfully
    """

    """
    get_xml_rpc is where the XML RPC connection to OpenNebula is established,
    we return our MockServer which is passed the value we want our XML RPC
    call to return.  We also return the XML RPC username & password
    """
    with patch("salt.cloud.clouds.opennebula._get_xml_rpc", MagicMock(return_value=[MockServer(["True","1","0"]), "username", "password"])):
        """
        get_vm_id gets called in the init routine, so we need to capture it
        """
        with patch("salt.cloud.clouds.opennebula.get_vm_id", MagicMock(return_value=1)):
            expected = {
                "action": "vm.disk_resize",
                "resized": "True",
                "vm_id": "1",
                "error_code": "0",
            }
            ret = opennebula.vm_disk_resize(VM_NAME, kwargs={"disk_id": "0", "disk_size": "1024"}, call="action")
            assert expected == ret
````

**Review by Ch3LL — APPROVED** [review]

**Comment by Ch3LL** [comment]
Looks like there are pre-commit and lint test failures here that need to be resolved. Also please ensure to rebase as we have resolved some test failures recently.

**Comment by Ch3LL** [comment]
Just following up here, any progress?

**Comment by KarlAustin** [comment]
Closing this PR as salt-cloud looks to be getting moved to community extension.

**Pull request closed without merging**